Created "check_pre_upgrade" actor to group all modules

artmello opened the pull request:
- [x] Created "CheckOutput" schema that will contain all fields necessary for a check module output
- [x] Modified existing pre upgrade modules actors to use "CheckOutput" schema
- [x] Created "check_pre_upgrade" to group all pre upgrade modules actors
- [ ] Implement defined LeApp check output schema into "CheckOutput"

shaded-enmity commented on src/schemas/checkoutput.py:
```diff
@@ -0,0 +1,16 @@
+@registered_schema('1.0')
+class CheckEntry(Document):
+    check_id = StringField()
+    status = StringField()
```
AFAICT this is an `enum` value, can you also list the possible options here? ('success' and 'error' i guess)

shaded-enmity commented on src/actors/upgrade/common/verify_pkg_rhsigned/verify_pkg_rhsigned.py:
```diff
@@ -47,11 +45,17 @@
+
+if not_signed:
+    check_result = [{
+        'check_id': context,
```
I'd just include the actor name here, if we need to scope the check better I'd add another field to the output structure - perhaps use something like `check_actor = <actor_name>` and `check_action = <optional_value>`. 
If we mash too much information into a single string value, we, or others, are destined to reverse-engineer the contents of the value sooner or later.

shaded-enmity commented on src/actors/upgrade/check_output/check_html_output/check_html_output.py:
```diff
@@ -0,0 +1,46 @@
+
+import json
+import sys
+from jinja2 import Template
```
We'll need to vendor `jinja2` with this actor I guess, can you please check jinjas' dependencies (transitively) so that we can determine what we need to vendor?

bertinatto replied:
+1. 

IMO it would be better to generate the HTML in the client side. A few advantages that I see in this approach:

1. We don't have to vendor `jinja2` with the actor.
2. In the client, we have the option to add a new output types in a decoupled way, i.e., we don't need another actor to generate a XML or CSV report.
3. If we want to make refinements to the web client (Cockpit), we don't have to change the actor as well.

One disadvantage for this approach is that each client (e.g.., `leappctl` or `cockpit-ui`) needs its own report template. I think this is acceptable, though.

If we decide to go with this approach, I think it would be OK to work on that  in a separate PR.

pirat89 replied:
> One disadvantage for this approach is that each client (e.g.., leappctl or cockpit-ui) needs its own report template. I think this is acceptable, though.

I guess you mean templates on the client side and report actor will produce report in suitable format for another processing (e.g. json).

bertinatto replied:
Yup, that's right.

artmello commented:
@shaded-enmity thanks for reviewing. Latest commit does apply your suggestions to CheckOutput schema. About vendoring jinja, I am not sure how to handle that right now but dependencies would be:
- github.com/pallets/jinja (python2-jinja2.rpm)
- github.com/pallets/markupsafe (python2-markupsafe.rpm)

pirat89 commented on src/actors/upgrade/common/list_content_pkg/list_content_pkg.py:
```diff
@@ -43,13 +42,21 @@
+    check_out = [{'checks': check_result}]
+    out.update({keys['check_out']: check_out})
 
 print(json.dumps(out))
```
Just reminder for others that used schema is not final solution and in future it will be different. We can expect later something like:
```diff
...
+        'status': 'inspection'
+        'summary': 'Detected <lang> modules that are not owned by any package.',
+        'body': You have installed <lang> modules that are not owned by any package. These will not be handled during upgrade. List of the untracked modules: {{ html_list .params }}
...
```
(that's really rough example)

bertinatto commented on src/actors/upgrade/common/list_content_pkg/_actor.yaml:
```diff
@@ -9,9 +9,9 @@ outputs:
-    name: ErrorList
+    name: CheckOutput
 description: |
   List which RPM package provides content provided as input or add an error message
```
Is this bit still valid? 

`(...)or add an error message`

bertinatto commented on src/actors/upgrade/common/list_content_pkg/list_content_pkg.py:
```diff
@@ -5,12 +5,11 @@
-    'err_ctx': 'context',
-    'err_value': 'value'
+    'check_out': 'check_output'
 }
```
Just for future reference, this isn't related to this PR (i.e., no need to change anything here):

IMO keeping this data structure to reference keys makes it more difficult to follow the code (even though it makes easier to change the keys). For instance, I guess this would a more readable way of getting `content`:

```context = ','.join(x['value'] for x in inputs.get('context', []))```

Anyway, this is just personal preference.

bertinatto commented on src/actors/upgrade/common/verify_pkg_rhsigned/verify_pkg_rhsigned.py:
```diff
@@ -47,11 +46,18 @@
+    print(json.dumps({keys['check_out']: check_out}))
+
+else:
+    print(json.dumps({}))
```
Won't this fail the schema validation?

pirat89 replied:
@bertinatto: nope. It will not.

bocekm commented on src/actors/upgrade/check_output/check_html_output/check_html_output.py:
```diff
@@ -0,0 +1,47 @@
+  <table style="width:100%">
+    {% for checks in check_output %}
+      {% for check in checks['checks'] %}
+        {% for param in check['params'] %}
```
The param is not required field in the CheckOutput schema but if it's missing in the message, nothing gets to the html report due to this for cycle.

artmello commented:
@bertinatto thanks for reviewing, I fixed the description and applied your suggestions to improve Python code. As mentioned that empty dict does not seem to break schema validation. Let me know what do you think.

@bocekm thanks for reviewing, too. I changed HTML output actor to still display output if 'param' is empty. Let me know how it goes for you

bertinatto commented:
Thanks, @artmello. I skimmed over the code again and it LGTM.

Just a note for our next iterations on this actor: IMO we should move the HTML generation to the clients (as per my comment above).
